### ISSUE TITLE
"Mode switch" for ansible-pull.

### ISSUE THREAD

**Issue by ccgthree** [ansible-community/ara#335]
Would it be possible to integrate a "mode switch" for ansible-pull that would replace the `/hosts` page with a page that instead reports the "controllers" as the hosts? It seems like it would be possible. I'm poking through the code right now to see if I can figure it out, but I thought I'd ask the masters while I tinker as a padawan. 

In my lab, I'm tricking the hosts report by using the FQDN of the node, rather than "localhost" as the target in the Ansible `hosts` configuration file. This works in a small-scale environment, but it'd be great if the sysadmin could just throw one switch on the ARA server that reconfigures the hosts page. 

Thank you so much for taking the time to make this useful tool. It's pretty amazing.

**Comment by dmsimard** [comment]
Hi @ccgthree and thanks for the issue.

I think it's a good idea and makes sense to be able to record the FQDN instead of localhost in use cases with ansible-pull.

> Would it be possible to integrate a "mode switch" for ansible-pull that would replace the /hosts page with a page that instead reports the "controllers" as the hosts? It seems like it would be possible.

The frontend is currently rather simple and doesn't manipulate or transform data: it takes what is provided by the API and then renders what's recorded in the database for humans.

This means we would need to have the callback record it correctly in the first place and then there is nothing special to do from a frontend perspective.

Currently the FQDN is recovered here: https://github.com/ansible-community/ara/blob/608006cec32d37878b197c0913c8c20f0f5c77b6/ara/plugins/callback/ara_default.py#L298

Hosts themselves are created with this method that is used where necessary: https://github.com/ansible-community/ara/blob/608006cec32d37878b197c0913c8c20f0f5c77b6/ara/plugins/callback/ara_default.py#L494-L499

I need to think about how this might work in practice and wonder if there is a way for the callback to know that it is running from ``ansible-pull`` instead of a regular ``ansible-playbook`` command. This also made me realize that although I know it works, we don't have integration tests that run ansible-pull.

I will try to experiment and report back.

> Thank you so much for taking the time to make this useful tool. It's pretty amazing.

Thank you :heart:

**Comment by dmsimard** [comment]
> I need to think about how this might work in practice and wonder if there is a way for the callback to know that it is running from ansible-pull instead of a regular ansible-playbook command.

I've experimented a bit and found that the most reliable way to do that is to determine if the parent process is ansible-pull since ansible-pull [shells out to ansible-playbook](https://github.com/ansible/ansible/blob/f42ffe16d29f00cca38b7b53566e18d36edfc461/lib/ansible/cli/pull.py#L263).

For example, I've tested that the following works well:
```
# Determine if the playbook is running from ansible-pull
parent = psutil.Process(os.getppid())
if parent.name() == "ansible-pull":
    self.log.debug("Playbook is running under ansible-pull")
```

Taking a step back, however, it makes me think that it's not just ansible-pull who may end up targetting localhost -- in other words: users of regular ansible-playbook commands may also want to flip localhost to the actual FQDN.

A potential way to address this kind of use case would be add a boolean callback configuration option like ``ARA_LOCALHOST_AS_FQDN`` that would do the flip when enabled.

**Comment by dmsimard** [comment]
o/ @ccgthree

I've proposed a PR with the ``ARA_LOCALHOST_AS_FQDN`` approach, I think it's simple enough and it works: https://github.com/ansible-community/ara/pull/336

There's still some TO DOs before we can land it but it's available to test if you'd like.

**Comment by ccgthree** [comment]
Wow, David. That's quick work! I'm going to test this out locally as well. 

In my localized testing, I've found that reporting the FQDN in the inventory file does work and doesn't seem to report back any errors or prevent any plays from running. The only concern that I think it would present is if a devices FQDN were to change and there were a sudden mismatch in DNS and PRT records. The solution presented in this PR, I think, is the cleanest solution.

**Comment by dmsimard** [comment]
The PR for implementing this feature (https://github.com/ansible-community/ara/pull/336) has merged and will be released in 1.5.8.
